### PR TITLE
chore: start migration to typescript

### PR DESCRIPTION
**Pull request by DorianMaliszewski**
## Type

- Migration

#### What is expected?

To typescript :

- ActionBar
- ActivityIndicator
- Typography
- Tooltip

#### The following changes where made:

1. Migrate proptypes to typescript interface and change defaultprops by function default param

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -5,7 +5,8 @@
     "jest": true
   },
   "parser": "@babel/eslint-parser",
-  "extends": ["@scaleway/react", "plugin:mdx/recommended"],
+  // TODO: Remove eslint-config-airbnb-typescript dependency when all the project is in typescript
+  "extends": ["@scaleway/react", "plugin:mdx/recommended", "airbnb-typescript"],
```
Comment (chambo-e) [inline .eslintrc:9]: Already included in @scaleway/eslint-config/typescript :)

File: src/theme/colors.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+// System colors
+const primary = violet
+const beta = orange
+const success = green
+const warning = red
+const info = blue
+
+const transparent = 'transparent'
+
+const colors: Record<string, string> = {
```
Comment (chambo-e) [inline src/theme/colors.ts:40]: Why ? You shouldn't to change this file no ? Just change the extension and it should just work ? 🤔

**Review by chambo-e — CHANGES_REQUESTED** [review]

File: .eslintrc [hunk 3]
```diff
@@ -52,7 +52,11 @@
       "parserOptions": {
         "project": ["tsconfig.json"]
       },
-      "extends": ["@scaleway/react/typescript"]
+      "extends": ["@scaleway/react/typescript"],
+      "rules": {
+        "react/require-default-props": "off",
```
Comment (chambo-e) [inline .eslintrc:57]: Rather https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/require-default-props.md#ignorefunctionalcomponents :)

File: src/components/Tooltip/index.tsx [hunk 4]
```diff
@@ -9,28 +9,54 @@ import {
...
+  | 'bottom-end'
+  | 'bottom'
+  | 'bottom-start'
+  | 'left-end'
+  | 'left'
+  | 'left-start'
+
 const variants = {
-  black: ({ theme }) => css`
+  black: ({ theme }: { theme: SCWUITheme }) => css`
```
Comment (chambo-e) [inline src/components/Tooltip/index.tsx:33]: We should use Emotion.Theme here no ?

File: src/components/Typography/__tests__/index.test.tsx [hunk 5]
```diff
@@ -1,22 +1,27 @@
...
-    ReactDOM.createPortal = jest.fn(element => element)
+    ;(ReactDOM.createPortal as unknown) = jest.fn(
+      element => element as unknown,
+    ) as unknown
   })
 
   afterEach(() => {
-    ReactDOM.createPortal.mockClear()
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-call
+    ;(ReactDOM.createPortal as unknown as { mockClear: () => void }).mockClear()
```
Comment (chambo-e) [inline src/components/Typography/__tests__/index.test.tsx:24]:
```suggestion
    (ReactDOM.createPortal as jest.Mock).mockClear()
```

File: src/components/Tooltip/index.tsx (hunk 4 above)
Comment (DorianMaliszewski) [reply src/components/Tooltip/index.tsx:33]: Ok :)

File: src/components/Typography/__tests__/index.test.tsx [hunk 6]
```diff
@@ -1,22 +1,26 @@
...
 import ReactDOM from 'react-dom'
 import Typography, { typographyVariants } from '..'
 import shouldMatchEmotionSnapshot from '../../../helpers/shouldMatchEmotionSnapshot'
 
+// TODO: Remove any but IDK what we can put here
 jest.mock(
   '../../Tooltip',
   () =>
-    ({ children, ...props }) =>
+    ({ children, ...props }: { children: (props: unknown) => unknown }) =>
```
Comment (matthprost) [inline src/components/Typography/__tests__/index.test.tsx:11]: What about `props: { [x: string]: unknown }` ? It's a bit same but at least it precise the structure

File: src/components/Typography/index.tsx [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+import PropTypes from 'prop-types'
+import React, { FocusEvent, MouseEvent, forwardRef, useCallback } from 'react'
+import recursivelyGetChildrenString from '../../helpers/recursivelyGetChildrenString'
+import { Color } from '../../theme/colors'
+import Box from '../Box'
+import Tooltip, { TooltipProps } from '../Tooltip'
+
+type StyleProps = {
+  theme?: Theme
+  maxLines?: string | number
```
Comment (matthprost) [inline src/components/Typography/index.tsx:12]: `maxLine` is defined as a number everywhere else in this file

```suggestion
  maxLines?: number
```

File: src/components/Typography/index.tsx [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+  children: PropTypes.node.isRequired,
+  color: PropTypes.string,
+  ellipsis: PropTypes.bool,
+  fontWeight: PropTypes.number,
+  lineHeight: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
+  maxLines: PropTypes.number,
+  onFocus: PropTypes.func,
+  onMouseEnter: PropTypes.func,
+  tooltipProps: PropTypes.shape({}),
+  variant: PropTypes.oneOf(typographyVariants),
```
Comment (matthprost) [inline src/components/Typography/index.tsx:343]: There is `width?: string | number` for type definition but the is none in propTypes

**Review by matthprost — CHANGES_REQUESTED** [review]

File: src/components/Typography/index.tsx [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+      `
+    : css`
+        ${styles[variant]?.({ theme } as StyleProps)}
+      `
+
+type StyledTextProps = {
+  ellipsis?: boolean
+  maxLines?: number
+  variant: TypographyVariant
+  as?: unknown
```
Comment (matthprost) [inline src/components/Typography/index.tsx:204]: Can't it be a string type ?

**Review by matthprost — CHANGES_REQUESTED** [review]

**Review by matthprost — APPROVED** [review]

**Review by chambo-e — APPROVED** [review]

**Pull request merged**